[Timestamps] Update for New Dashboard (Take 2)

motackt opened the pull request:
This gets timestamps up and running on the new Dashboard, and updates the loading icon so it works on other colour palettes too.

(A clean version of #1843).

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -230,6 +265,89 @@ XKit.extensions.timestamps = new Object({
+			}
+
+			var xtimestamp_html = `<div class="xkit_timestamp_${post_id} ${xtimestamp_class} xtimestamp_loading">&nbsp;</div>`;
+			post.find(XKit.extensions.timestamps.meatballs_class).prepend(xtimestamp_html);
```
nice try, but we should leave the header alone and replicate the timestamps behaviour that everyone's used to. the timestamps should be inserted into the post after the `header` element.

when creating this new block element, you'll want to use the CSS variable `--post-padding` for the left/right padding, `--post-header-vertical-padding` for the bottom margin, and force the element's height to exist with `line-height: 1em`. some negative top margin would look nice too

my mockup:
![image](https://user-images.githubusercontent.com/28949509/83654420-00d76080-a5b5-11ea-9d14-97f5f291db51.png)

CSS used:
```css
font-size: 14px;
line-height: 1em;
padding: 0 var(--post-padding);
margin-top: calc(var(--post-header-vertical-padding) / -2);
margin-bottom: var(--post-header-vertical-padding);
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -230,6 +265,89 @@ XKit.extensions.timestamps = new Object({
+	},
+
+	react_fetch_timestamp: async function(post_id, date_element) {
+		if (this.fetch_from_cache(post_id, date_element)) {
+			return;
+		}
```
the cache is unnecessary now that timestamps are included in the post props - there's no additional network requests to save on when getting them

AprilSylph replied:
if we're still using `fetch_timestamp` for reblog timestamps then this is indeed unnecessary

motackt replied:
Wait, why shouldn't we fetch from cache when using the regular fetch_timestamp for reblogs?

AprilSylph replied:
we should, but this is in `react_fetch_timestamp`. [see this conversation in the Files tab](https://github.com/new-xkit/XKit/pull/1844/files/3c6fe29c8d9a43f17afeaf21285d84c357b4b9f3#diff-13325f4522cfe04acaed335ee61812f9R335)

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -245,14 +363,15 @@ XKit.extensions.timestamps = new Object({
 	show_failed: function(obj) {
 		$(obj).html("failed to load timestamp <div class=\"xkit-timestamp-failed-why\">why?</div>");
-		$(obj).removeClass('xtimestamp_loading');
+		if ($(obj).hasClass('xtimestamp_loading')) { $(obj).removeClass('xtimestamp_loading'); }
```
redundant; `.removeClass` already tests for classes before removing them, or otherwise does nothing if the class to be removed isn't there

```suggestion
		$(obj).removeClass('xtimestamp_loading');
```

motackt commented on Extensions/timestamps.js:
```diff
@@ -122,23 +121,58 @@ XKit.extensions.timestamps = new Object({
+				}
+			});
+
+			return;
+
+		} else {
```
Agree on the redundancy, but I feel the else block makes it more clear to future programmers that the following is only meant to run on the old Dashboard. Can't resolve by removing that return because it makes the program break. Either or, I'm happy to remove the else block to keep the code consistent of course.

AprilSylph replied:
huh? how does removing the return break it when there's no other code in the function after the else block?

(also in favour of removing the else block for consistency)

motackt replied:
I thought it was weird too, XKit wouldn't even load the extension, but I'm unable to reproduce it now so I guess whatever wasn't working is fixed now ¯\_(ツ)_/¯

No worries, removed the else block!

motackt commented on Extensions/timestamps.js:
```diff
@@ -122,23 +121,58 @@ XKit.extensions.timestamps = new Object({
+				this.meatballs_class = rightContent.map(x => `.${x}`).join(", ");
+				this.reblogs_class = reblog.map(x => `.${x}`).join(", ");
+				this.reblog_headers_class = reblogHeader.map(x => `.${x}`).join(", ");
+				this.reblog_op_class = badge[1];
```
Tried to fix this by removing this line, and in `react_add_timestamps()` we select all posts, then get the elements with the reblog class, then pick just the first one.

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -122,23 +121,58 @@ XKit.extensions.timestamps = new Object({
+				return await tumblr.getCssMap();
+			})
+			.then(({post, rightContent, reblog, reblogHeader, badge, blogLink, private}) => {
+				this.posts_class = post.map(x => `.${x}`).join(", ");
+				this.meatballs_class = rightContent.map(x => `.${x}`).join(", ");
+				this.reblogs_class = reblog.map(x => `.${x}`).join(", ");
+				this.reblog_headers_class = reblogHeader.map(x => `.${x}`).join(", ");
+				this.reblog_op_class = badge[1];
+				this.blog_link_class = blogLink.map(x => `.${x}`).join(", ");
+				this.private_posts_class = private.map(x => `.${x}`).join(", ");
```
1. make sure you're using `XKit.css_map.getCssMap()` first; otherwise, the other functions will always return undefined
2. `keyToCss` takes a string argument, while `blogLink` here is an array from the CSS map, which we don't want because we then have to transform it into a selector list.

```js
XKit.css_map.getCssMap().then(() => {
	this.blog_link_selector = XKit.css_map.keyToCss("blogLink");
	// this.blog_link_selector is now ".hVK3L, ._23-Al, ._1PyOv, ._2A1wD"
});
```

motackt replied:
Ahhh there we go, I knew I should be giving it strings of the keys but I didn't know I needed to remove the curly braces and the keys in there. Thanks!

AprilSylph replied:
that's not it, although the argument destructuring does become unnecessary.
your example used `XKit.css_map.keyToCss(blogLink)`, but `blogLink` is a variable name which points to an array. `XKit.css_map.keyToCss("blogLink")` is passing a string in directly

motackt replied:
Way too scatterbrained for programming, I'm unable to reproduce this error as well XD

AprilSylph commented on Extensions/timestamps.css:
```diff
@@ -30,17 +30,26 @@
+}
+
+.xkit--react .xtimestamp {
+	top: 0;
```
probably better to just make the `top` rule not do anything; also, `color` needs to be overridden

```suggestion
	position: static;
	color: var(--black);
```

motackt commented:
I feel like --gray-40 is a better fit for the timestamp colour, because it's meant to be more of a secondary element after the important header information.

Thoughts?

AprilSylph commented:
gray-40 is a little hard to read - let's match the tags color for consistency and better readability (gray-65)

motackt commented:
Agreed. I'll push that commit in a second after I investigate this bug where reblog *names* will be replaced with the "failed to load timestamp why?" text.

AprilSylph commented on Extensions/timestamps.css:
```diff
@@ -39,13 +39,28 @@
+	line-height: 1em;
+	padding: 0 var(--post-padding);
+	margin-left: 0;
+	margin-top: calc(-1 * (var(--post-header-vertical-padding) / 2));
```
?

does dividing by -2 not work on chrome?

motackt replied:
It does, I thought this would be a little more readable, but I can change it back to `/ -2` no problem.

motackt commented:
Ready for review ^__^

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +253,88 @@ XKit.extensions.timestamps = new Object({
 	},
 
+	react_add_timestamps: function() {
+		var posts = $("[data-id]:not(.xkit_timestamps)");
```
probably best to immediately add the `xkit_timestamps` class once we've collected them for processing
```suggestion
		var posts = $("[data-id]:not(.xkit_timestamps)");
		posts.addClass("xkit_timestamps");
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +253,88 @@ XKit.extensions.timestamps = new Object({
+
+		posts.each(function() {
+			var post = $(this);
+			post.addClass("xkit_timestamps");
```
(and then remove the redundant per-post class-adding here)
```suggestion
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +253,88 @@ XKit.extensions.timestamps = new Object({
+			var post = $(this);
+			post.addClass("xkit_timestamps");
+
+			var post_id = $(this).attr("data-id");
```
`$(this)` is already saved as `post`; this is significant because `$` is a function, so it's best to reduce calls to it where possible for efficiency

```suggestion
			var post_id = post.attr("data-id");
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +253,88 @@ XKit.extensions.timestamps = new Object({
+			}
+
+			var xtimestamp_html = `<div class="xkit_timestamp_${post_id} ${xtimestamp_class} xtimestamp-loading">&nbsp;</div>`;
+			$(xtimestamp_html).insertAfter(post.find(XKit.extensions.timestamps.headers_class));
```
can be simplified

```suggestion
			$(xtimestamp_html).insertAfter(post.find("header"));
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +253,88 @@ XKit.extensions.timestamps = new Object({
+			var post_id = $(this).attr("data-id");
+
+			var xtimestamp_class = "xtimestamp";
+			if (XKit.extensions.timestamps.in_search && !$("#search_posts").hasClass("posts_view_list")) {
+				xtimestamp_class = "xtimestamp-in-search";
+			}
```
does `#search_posts` or the `.posts_view_list` class even exist on the react search page?

motackt replied:
I definitely can't find it, I'll just remove that part.

AprilSylph commented:
i'm pretty sure `react_add_reblog_timestamps()` should be rewritten using the `trail` array of objects in each post's properties, which map directly to the reblog elements in a post. in particular, the `blog.uuid` and `post.id` of each `trail` object is a much clearer and reliable method of getting the blog identifier and post ID than trying to parse them out of the HTML. we should also be using `tumblr.apiFetch` to grab the timestamps.

reference: [x8/src/reblog_timestamps.js](https://github.com/new-xkit/x8/blob/master/src/reblog_timestamps.js)

(this is a lot of work: i'll help with it once there's no other changes needed!)

motackt commented:
I hadn't noticed that the post props gave a trail! That's very helpful, and yes would be much better than trying to parse important details from the HTML. Maybe should be something for another Pull Request though just to get Timestamps out the door faster :?

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -108,37 +107,58 @@ XKit.extensions.timestamps = new Object({
+				this.blog_link_class = XKit.css_map.keyToCss("blogLink");
+			});
+
+			if (this.preferences.posts.value || (this.preferences.inbox.value && XKit.interface.where().inbox)) {
```
the inbox doesn't have a react page yet, so this second condition will never activate/is redundant for now

```suggestion
			if (this.preferences.posts.value) {
```

motackt replied:
Shouldn't we keep this here for when the react dash does get an inbox? Less work for the future, right?

AprilSylph replied:
i guess so. my line of thinking was that we don't know if `react_add_timestamps()` will work on the react inbox, but if it doesn't or produces erroneous behaviour, then that still applies to everyone that has all post timestamps enabled anyway...

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -108,37 +107,58 @@ XKit.extensions.timestamps = new Object({
+			}
+
+			if (this.preferences.only_on_hover.value) {
+				XKit.tools.add_css(`.xtimestamp { display: none; } ${this.posts_class.split(", ").map(x => x + ":hover .xtimestamp").join(", ")} { display: block; }`, "timestamps_on_hover");
```
`this.posts_class` isn't guaranteed to be defined here. 

`XKit.css_map.getCssMap()` is an asynchronous function, which means that technically, it (and any functions attached with `.then()`) may not have finished executing by the time the next line, or any of the rest of the code block executes.

all the code in the `if (XKit.page.react)` block (currently below the `.then()` function) should be in the `.then()` function to guarantee that the necessary values have been set.

motackt replied:
That's a great point, forgot about that

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -195,7 +187,7 @@ XKit.extensions.timestamps = new Object({
 			var blog_name = post.attr('data-tumblelog-name');
 
-			if (XKit.extensions.timestamps.in_search && !$("#search_posts").hasClass("posts_view_list")) {
+			if (XKit.extensions.timestamps.in_search) {
```
i assume `#search_posts` still exists on the old dash; this shouldn't be touched
```suggestion
			if (XKit.extensions.timestamps.in_search && !$("#search_posts").hasClass("posts_view_list")) {
```

motackt replied:
This is why u don't code rly early in the morning XD

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +245,91 @@ XKit.extensions.timestamps = new Object({
+			return;
+		}
+
+		XKit.extensions.timestamps.check_quota();
```
the cache is (or should be) unused for post timestamps, but should be used for reblog timestamps since they have to be fetched; the quota doesn't need to be checked here

```suggestion
```

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +245,91 @@ XKit.extensions.timestamps = new Object({
+		.each(function() {
+			var reblog = $(this);
+
+			if (reblog.length === 0) { return; }
```
when would `.each()` pass in a nonexistent element?

motackt replied:
Ah, ur suggesting an early return like `if (reblogs.length === 0) { return; }`, right?

AprilSylph replied:
if `reblogs` is an empty set, then both `.addClass` and `.each` just won't do anything, so it's redundant

AprilSylph commented on Extensions/timestamps.js:
```diff
@@ -224,12 +245,91 @@ XKit.extensions.timestamps = new Object({
+			reblog.find(XKit.extensions.timestamps.reblog_headers_class).append(normal_html);
+			var note = $(`.xkit_timestamp_${post_id}`);
+
+			// XKit.interface.react.post_props doesn't appear to work for reblogs yet
+//			XKit.extensions.timestamps.react_fetch_timestamp(post_id, note);
```
unnecessary comments; `post_props` will never have reblog timestamps, since they aren't saved as part of the trail

```suggestion
```

motackt replied:
):

motackt commented:
Added the destroy method and fixed the parsing for original reblogs (forgot that when looking for posts using [data-id], you're further up the tree than when you're using the CSS Map's posts class).

AprilSylph commented:
well, since i've already fixed `async_add_function` in another PR, i'm just gonna shuffle my function into this PR before merging it
